editor.planx.uk/src/pages/FlowEditor/components/EditHistory.tsx

```diff
@@ -0,0 +1,223 @@
+import { gql, useSubscription } from "@apollo/client";
+import RestoreOutlined from "@mui/icons-material/RestoreOutlined";
+import Timeline from "@mui/lab/Timeline";
+import TimelineConnector from "@mui/lab/TimelineConnector";
+import TimelineContent from "@mui/lab/TimelineContent";
+import TimelineDot from "@mui/lab/TimelineDot";
+import TimelineItem, { timelineItemClasses } from "@mui/lab/TimelineItem";
+import TimelineSeparator from "@mui/lab/TimelineSeparator";
+import Box from "@mui/material/Box";
+import IconButton from "@mui/material/IconButton";
+import Typography from "@mui/material/Typography";
+import SimpleExpand from "@planx/components/shared/Preview/SimpleExpand";
+import { formatOps } from "@planx/graph";
+import { OT } from "@planx/graph/types";
+import DelayedLoadingIndicator from "components/DelayedLoadingIndicator";
+import React, { useState } from "react";
+
+import { formatLastEditDate, Operation } from "..";
+import { useStore } from "../lib/store";
+
+const EditHistory = () => {
+  const OPS_TO_DISPLAY = 5;
+
+  const [focusedOpIndex, setFocusedOpIndex] = useState<number | undefined>(
+    undefined,
+  );
+
+  const [flowId, flow, canUserEditTeam, teamSlug, undoOperation] = useStore(
+    (state) => [
+      state.id,
+      state.flow,
+      state.canUserEditTeam,
+      state.teamSlug,
+      state.undoOperation,
+    ],
+  );
+
+  const { data, loading, error } = useSubscription<{ operations: Operation[] }>(
+    gql`
+      subscription GetRecentOperations($flow_id: uuid = "") {
+        operations(
+          limit: 15
+          where: { flow_id: { _eq: $flow_id } }
+          order_by: { created_at: desc }
+        ) {
+          id
+          createdAt: created_at
+          actor {
+            id
+            firstName: first_name
+            lastName: last_name
+          }
+          data(path: "op")
+        }
+      }
+    `,
+    {
+      variables: {
+        flow_id: flowId,
+      },
+    },
+  );
```

Comment:
I spent a bit of time thinking about what how we might include "publishes" alongside "operations" in the edit history timeline and there's a few tricky details to work out / good reasons to re-visit this post-MVP I think:
- Unlike `query` or `mutation`, GraphQL's `subscription` can have exactly one root object (meaning I can't subscribe to `operations` & `published_flows` in the same hook
- Defining in a meaningful `published_flows` subscription hook is a bit tricky - the `where` condition we really want is matching flow id _and_ published date is greater than or equal to the oldest operation that we fetched (basically, simple "last 15" like operations won't make sense)
- We wouldn't be able to display the "restore" icon/action next to "Published by..." items in the history list, but they could serve as static context markers to help understand if you're undo-ing operations to a state _before_ your last publish

Reply:
Thanks for the write up here.

I was also thinking about this and couldn't think of a good way to communicate that you can restore operations to a previous version before your last publishing event, but that's not going to undo the publish event - the old content will still be live.

I actually think if there's a user-need for this, it might be a separate thing (in another tab?) called "Publishing history".

Support leaving this as-is for now 👍